zerver/tests/test_message_fetch.py

```diff
@@ -1880,6 +1970,95 @@ def test_get_messages_with_narrow_stream(self) -> None:
                 self.assertEqual(message["type"], "stream")
                 self.assertEqual(message["recipient_id"], stream_recipient_id)
 
+    def test_get_messages_with_narrow_streams_subscribed(self) -> None:
+        stream_dicts: List[StreamDict] = [
+            {
+                "name": "public_stream",
+                "description": "Public stream with public history",
+            },
+            {
+                "name": "private_stream",
+                "description": "Private stream with non-public history",
+                "invite_only": True,
+            },
+            {
+                "name": "private_with_history",
+                "description": "Private stream with public history",
+                "invite_only": True,
+                "history_public_to_subscribers": True,
+            },
+        ]
+
+        realm = get_realm("zulip")
+        created, existing = create_streams_if_needed(realm, stream_dicts)
+        self.assertEqual(len(created), 3)
+        self.assertEqual(len(existing), 0)
+
+        cordelia = self.example_user("cordelia")
+        self.login("cordelia")
+        self.subscribe(cordelia, stream_dicts[1]["name"])
+        self.subscribe(cordelia, stream_dicts[2]["name"])
+
+        self.login("hamlet")
+        hamlet = self.example_user("hamlet")
+
+        last_msg_id = self.get_last_message().id + 1
+
+        for stream in stream_dicts:
+            self.send_stream_message(
+                sender=cordelia,
+                stream_name=stream["name"],
+            )
+
+        self._update_tsvector_index()
+
+        narrow = [dict(operator="streams", operand="subscribed")]
+
+        result: Dict[str, Any] = self.get_and_check_messages(
+            dict(
+                narrow=orjson.dumps(narrow).decode(),
+                anchor=last_msg_id,
+                num_before=0,
+                num_after=10,
+            )
+        )
+
+        # Before the user subscribes no message in narrow
+        self.assertEqual(len(result["messages"]), 0)
+
+        self.subscribe(hamlet, stream_dicts[0]["name"])
+        self.subscribe(hamlet, stream_dicts[1]["name"])
+        self.subscribe(hamlet, stream_dicts[2]["name"])
+
+        result = self.get_and_check_messages(
+            dict(
+                narrow=orjson.dumps(narrow).decode(),
+                anchor=last_msg_id,
+                num_before=0,
+                num_after=10,
+            )
+        )
+
+        # Messages will be 2 not 3 first due to the subscribed
+        # public stream and second due to the private stream
+        # with shared history. Private stream with unshared
+        # history does not contribute to message in narrow.
+        self.assertEqual(len(result["messages"]), 2)
```

Comment:
It's often better to check the specific message IDs rather than a count, as that's more robust should we need to extend the test, as well as problems where we exclude the wrong message.